### PR TITLE
Adding 9 regression tests for fam context modelling

### PR DESCRIPTION
**Pull request by cynara-justine**
Adding 9 regression tests for fam context modelling
Also added a temporary fix for throwing error in some cases in fam_context.h file to prevent the issue where earlier it was silently proceeding ignoring the error

### CODE REVIEW

File: test/reg-test/fam-api-reg/fam_context_reg_test.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+{
+Fam_Region_Descriptor *rd = NULL;
+Fam_Descriptor *descriptor = NULL;
+fam_context *ctx;
+for (int i=0; i<NUM_IO_ITERATIONS; i++)
+    { 
+EXPECT_NO_THROW(ctx= my_fam->fam_context_open());
+
+        // create a 100 MB region with 0777 permissions
+        EXPECT_NO_THROW(rd= my_fam->fam_create_region(
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_context_reg_test.cpp:121]: Can you move this create region outside of the loop

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: test/reg-test/fam-api-reg/fam_context_reg_test.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        EXPECT_NO_THROW(ctx->fam_or(descriptor, 0, (uint32_t)2));
+        EXPECT_NO_THROW(ctx->fam_or(descriptor, 0, (uint32_t)3));
+        EXPECT_NO_THROW(ctx->fam_or(descriptor, 0, (uint32_t)4));
+        EXPECT_NO_THROW(my_fam->fam_or(descriptor, 0, (uint32_t)4));
+        EXPECT_NO_THROW(ctx->fam_quiet());
+        EXPECT_NO_THROW(my_fam->fam_quiet());
+
+        // We are done with the operations. Destroy the region and everything in
+        // it
+        EXPECT_NO_THROW(my_fam->fam_destroy_region(rd));
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_context_reg_test.cpp:125]: This should be outside the loop

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by faizan-barmawer — APPROVED** [review]
These test cases look fine.
We can add more test scenario:
1. Cover all I/O APIs - you have covered only atomic APIs, add get/put/scatter/gather blocking and non-blocking
2. Use same context from multiple threads, doing different operations - get, put, progress, etc.
3. Add few negative tests - call control path APIs like allocate/create region from context object. This should throw exception.

You can add more tests in this PR or can raise a new PR, anything is fine.

Regards,
Faizan

**Comment by cynara-justine** [comment]
Thanks for reviewing.  Please go ahead with merging these changes.  I shall add new test cases and raise a new PR for those.

Thanks and regards,
Cynara

File: test/reg-test/fam-api-reg/fam_context_mt_reg_test.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    fam_context *ctx;
+    EXPECT_NO_THROW(ctx = my_fam->fam_context_open());
+    EXPECT_NO_THROW(ctx->fam_or(item, 0, (uint32_t)1));
+    EXPECT_NO_THROW(ctx->fam_or(item, 0, (uint32_t)2));
+    EXPECT_NO_THROW(ctx->fam_or(item, 0, (uint32_t)3));
+    EXPECT_NO_THROW(ctx->fam_or(item, 0, (uint32_t)4));
+    EXPECT_NO_THROW(my_fam->fam_or(item, 0, (uint32_t)4));
+    EXPECT_NO_THROW(ctx->fam_quiet());
+    EXPECT_NO_THROW(my_fam->fam_quiet());
+    // We are done with the operations. Destroy the region and everything in it
```
Comment (mashoodk) [inline test/reg-test/fam-api-reg/fam_context_mt_reg_test.cpp:94]: Can you remove second part of the comment. We are not destroying the region here.

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by mashoodk — APPROVED** [review]

**Pull request merged**